core/templates/dev/head/pages/exploration_editor/editor_tab/SolutionVerificationServiceSpec.js

```diff
@@ -17,183 +17,204 @@
+    var interactionCtrl = $controller('StateInteraction', {
+      $scope: scope,
+      editorContextService: ecs,
+      changeListService: cls,
+      explorationStatesService: ess,
+      editabilityService: {
+        isEditable: function() {
+          return true;
+        }
+      },
+      stateInteractionIdService: siis,
+      stateCustomizationArgsService: scas,
+      interactionDetailsCache: idc,
+      INTERACTION_SPECS: IS
+    });
+  }));
 
-      var interactionCtrl = $controller('StateInteraction', {
-        $scope: scope,
-        editorContextService: ecs,
-        changeListService: cls,
-        explorationStatesService: ess,
-        editabilityService: {
-          isEditable: function() {
-            return true;
-          }
+  it('should verify a correct solution',
+    function(done) {
+      ecs.setActiveStateName('First State');
+      scope.initStateEditor();
+      var state = ess.getState('First State');
+      siis.init(
+        'First State', state.interaction.id, state.interaction, 'widget_id');
+      scas.init(
+        'First State', state.interaction.customizationArgs,
+        state.interaction, 'widget_customization_args');
+
+      siis.displayed = 'TextInput';
+      scope.onCustomizationModalSavePostHook();
+
+      ess.saveSolution('First State', sof.createNew(false, 'abc', 'nothing'));
+
+      svs.verifySolution(0, state,
+        ess.getState('First State').interaction.solution.correctAnswer,
+        function() {
+          ess.updateSolutionValidity('First State', true);
         },
-        stateInteractionIdService: siis,
-        stateCustomizationArgsService: scas,
-        interactionDetailsCache: idc,
-        INTERACTION_SPECS: IS
-      });
-    }));
-
-    it('should verify the solution correctly',
-      function() {
-        ecs.setActiveStateName('First State');
-        scope.initStateEditor();
-
-        var state = ess.getState('First State');
-        siis.init(
-          'First State', state.interaction.id, state.interaction, 'widget_id');
-        scas.init(
-          'First State', state.interaction.customizationArgs,
-          state.interaction, 'widget_customization_args');
-
-        siis.displayed = 'TextInput';
-        scope.onCustomizationModalSavePostHook();
-
-        ess.saveSolution('First State', sof.createNew(false, 'abc', 'nothing'));
-
-        svs.verifySolution(0, state,
-          ess.getState('First State').interaction.solution.correctAnswer,
-          function() {
-            ess.updateSolutionValidity('First State', true);
-            expect(ess.isSolutionValid('First State')).toBe(true);
-          },
-          function() {
-            ess.updateSolutionValidity('First State', false);
-          });
+        function() {
+          ess.updateSolutionValidity('First State', false);
+        }
+        ).then(function() {
+          expect(ess.isSolutionValid('First State')).toBe(true);
+          done();
+        });
+      scope.$apply();
```

Comment:
Ah, I see. OK, I have a few suggestions for how to make it less flaky.

- This test actually reads a bit weirdly. You shouldn't need to call updateSolutionValidity; you just want to make sure that the right callback is called, right? So just create two dummy functions, spy on both of them, and check at the end that one has been called and the other hasn't -- e.g. http://arches.io/2012/12/testing-callbacks-with-jasmine-spies/
- Don't return anything from verifySolution(). Instead, use afterEach() for the check at the end, like [here](https://jasmine.github.io/2.0/introduction.html#section-Asynchronous_Support). That should guarantee that the check gets done only after at least one done() is called.